### ISSUE TITLE
Levenshtein slow and high on CPU usage

### ISSUE THREAD

**Issue by mishushakov** [crystal-lang/crystal#11335]
## Bug Report

Crystal 1.2.0 (2021-10-13)

LLVM: 11.1.0
Default target: x86_64-apple-macosx

Levenshtein is very slow and uses too much CPU when trying to match a large sample over a large text corpus

<img width="747" alt="Screenshot 2021-10-18 at 16 06 58" src="https://user-images.githubusercontent.com/10400064/137748115-3aea0f9e-892e-4216-9ea3-75ca10def6a3.png">

https://user-images.githubusercontent.com/10400064/137750280-fc1fd225-0a5c-44b6-b2e1-500a8b93198d.mov

it takes up to 5 minutes to find a match

Case: software license classifier

```crystal
license_file = File.open("src/licenses.csv")
sample = File.read("LICENSE")
license_db = CSV.parse(license_file)
license_db.each do |entry|
  puts Levenshtein.distance(sample, entry[1])
end
```

[licenses.csv](https://github.com/crystal-lang/crystal/files/7365700/licenses.csv)

**Comment by mishushakov** [comment]
here's another example, more similar to the node.js one

```crystal
CSV.each_row(@license_file) do |row|
  puts Levenshtein.distance(sample, row[1])
end
```

<img width="448" alt="Screenshot 2021-10-18 at 17 45 46" src="https://user-images.githubusercontent.com/10400064/137765098-f56d2efd-df48-4946-8495-cbdaeae12ad3.png">

**Comment by j8r** [comment]
Multi-threading is not enabled by default, `-Dpreview_mt` has to be put. And also creating so much fibers can be counter-productive. You can create, say 4 fibers, then dispatch each row to each of them through to each of the 4 channels.

**Comment by asterite** [comment]
@j8r I don't think multi-threading has anything to do here. The algorithm used by the node-js library is better, and we should improve ours too.

**Comment by mishushakov** [comment]
thanks for following up, @asterite 
the license file contains a MIT license template

the output was different, because the sample was a little bit different

**Comment by mishushakov** [comment]
correct, the first test was not on release mode
license checker is not a requirement for the idea i'm building and even the results of Node.JS were suboptimal

i need a better algorithm
thanks all for your attention :)

**Comment by jzakiya** [comment]
See if these make any difference.
https://rosettacode.org/wiki/Levenshtein_distance#Crystal

**Comment by vlazar** [comment]
Tried Crystal implementations from rosetta code https://rosettacode.org/wiki/Levenshtein_distance#Crystal

First one is terribly slow.
The second one is almost 7x slower than the one in standard library.

**Comment by asterite** [comment]
Yes, it uses a different algorithm. If you search for levensthein Myers you'll find it.

It actually only allocated that array once (js doesn't have threads). We could allocate that array on the stack, or once per thread.

We should definitely switch to the algorithm. I don't have time to do it, but it's a nice task for someone who liked these things. It doesn't matter if it's less readable than the current implementation, performance is more important.

**Comment by asterite** [comment]
It seems it's explained here: https://www.google.com/url?sa=t&source=web&rct=j&url=https://citeseerx.ist.psu.edu/viewdoc/download%3Fdoi%3D10.1.1.142.1245%26rep%3Drep1%26type%3Dpdf&ved=2ahUKEwi2g9Wq6djzAhVVqJUCHTDDBY0QFnoECAYQAQ&sqi=2&usg=AOvVaw0PHsNv91SBeHQg8REvhp2_

I might read it and implement it if I have time, but I can't promise anything.
